core/state_processor.go

```diff
@@ -168,22 +166,75 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 							return nil, nil, 0, fmt.Errorf("error decoding bytes %x: %w", stIt.Slot(), err)
 						}
 						copy(safeValue[32-len(value):], value)
-							slotnr := rawdb.ReadPreimage(statedb.Database().DiskDB(), stIt.Hash())
+						slotnr := rawdb.ReadPreimage(statedb.Database().DiskDB(), stIt.Hash())
 						fdb.LastSlotHash = stIt.Hash()
 
+						// move the storage to the next value, so that if the max
+						// leaf count is reached, the loop interrupted, and resumed
+						// at the next block, `stIt.Next()` will check fot the presence
+						// of the next slot, and mark the whole storage as "processed"
+						// if no such slot can be found.
+						for x := 31; x >= 0; x-- {
+							if fdb.LastSlotHash[x] < 255 {
+								break
+							}
+							fdb.LastSlotHash[x] = 0
+
+							// This means that 0xff....ff will circle back to 0x00..00
+							// but it's not a problem because the next call to stIt.Next()
+							// will update the processing marker, so the iterator will
+							// not be used again.
+						}
+
 						mkv.addStorageSlot(addr, slotnr, safeValue[:])
+
+						// advance the storage iterator and leave the loop
+						// if the iterator reached the end.
+						fdb.StorageProcessed = !stIt.Next()
+						if fdb.StorageProcessed {
+							break
 						}
+					}
 					stIt.Release()
 				}
+
+				// Process the account if all of its storage slots have been processed.
+				// If the maximum number of processable leafs has been reached, this
+				// part is skipped until the next block. If the limit was reached right
+				// after processing the storage, the account will be processed in the
+				// next block as the check for !fdb.StorageProcessed will be false then.
+				if fdb.StorageProcessed && count < maxMovedCount {
+					count++ // count increase for the account itself
+
+					mkv.addAccount(addr, acc)
+
+					// Store the account code if present
+					if !bytes.Equal(acc.CodeHash, emptyCodeHash[:]) {
+						code := rawdb.ReadCode(statedb.Database().DiskDB(), common.BytesToHash(acc.CodeHash))
+						chunks := trie.ChunkifyCode(code)
+						count += len(chunks) // count increase for the code chunks
+						// this means that more than maxMovedCount leaves can be
+						// transferred for a single block, but it's good enough
+						// for now.
+
+						mkv.addAccountCode(addr, uint64(len(code)), chunks)
+					}
+
+					// reset storage iterator marker for next account
+					fdb.LastSlotHash = common.Hash{}
+				}
 			}
```

Comment:
_I think_ there's a corner case that isn't correctly handled here; let me explain to double-check with you.

- Let's suppose we're at L206 and `fdb.StorageProcessed == true` and `count < maxMovedCount`. OK, we finished with the storage slots, and we also have room to migrate the account. We migrate the account so `count++`.
- We'll jump again to the main loop at L125 doing `for accIt.Next() && count < maxMovedCount`, but suppose that `count < maxMovedCount == false`. i.e the last account that we fully migrated fitted _perfectly_ into the limit.

In this situation, `fdb.LastAccHash` is still that _fully_ migrated account... so in the next block looks like we'll migrate it again. This sounds like a similar situation that (I think) you wanted to solve for storage-slots.

Reply:
right, initially I was doing like you said: not checking for `count < maxMovedCount`. I even have a comment to that effect. So you are correct that we need something similar, and I'm starting to see the wisdom in what you said before: revert and add a TODO.